pkg/controller/operators/catalog/step.go

```diff
@@ -98,61 +111,161 @@ func (b *builder) NewCRDStep(manifest string, client clientset.Interface, status
+		case v1alpha1.StepStatusCreated:
+			return v1alpha1.StepStatusCreated, nil
+		case v1alpha1.StepStatusWaitingForAPI:
+			crd, err := client.CustomResourceDefinitions().Get(context.TODO(), step.Resource.Name, metav1.GetOptions{})
+			if err != nil {
+				if k8serrors.IsNotFound(err) {
+					return v1alpha1.StepStatusNotPresent, nil
+				} else {
+					return v1alpha1.StepStatusNotPresent, errorwrap.Wrapf(err, "error finding the %s CRD", crd.Name)
+				}
+			}
+			established, namesAccepted := false, false
+			for _, cdt := range crd.Status.Conditions {
+				switch cdt.Type {
+				case apiextensionsv1beta1.Established:
+					if cdt.Status == apiextensionsv1beta1.ConditionTrue {
+						established = true
+					}
+				case apiextensionsv1beta1.NamesAccepted:
+					if cdt.Status == apiextensionsv1beta1.ConditionTrue {
+						namesAccepted = true
+					}
+				}
+			}
+			if established && namesAccepted {
+				return v1alpha1.StepStatusCreated, nil
+			}
+		case v1alpha1.StepStatusUnknown, v1alpha1.StepStatusNotPresent:
+			crd, err := crdlib.UnmarshalV1Beta1(step.Resource.Manifest)
+			if err != nil {
 				return v1alpha1.StepStatusUnknown, err
 			}
+			b.logger.Debugf("creating v1beta1 CRD %#v", crd)
+
+			_, createError := client.CustomResourceDefinitions().Create(context.TODO(), crd, metav1.CreateOptions{})
+			if k8serrors.IsAlreadyExists(createError) {
+				currentCRD, _ := client.CustomResourceDefinitions().Get(context.TODO(), crd.GetName(), metav1.GetOptions{})
+				// Compare 2 CRDs to see if it needs to be updated
+				logger.Debugf("\n current crd: %#v \n new crd: %#v \n", currentCRD, crd)
+				if crdlib.V1Beta1NotEqual(currentCRD, crd) {
+					b.logger.Debugf("not equal")
+					// Verify CRD ownership, only attempt to update if
+					// CRD has only one owner
+					// Example: provided=database.coreos.com/v1alpha1/EtcdCluster
+					matchedCSV, err := index.V1Beta1CRDProviderNames(b.csvToProvidedAPIs, crd)
+					if err != nil {
+						return v1alpha1.StepStatusUnknown, errorwrap.Wrapf(err, "error find matched CSV: %s", step.Resource.Name)
+					}
+					crd.SetResourceVersion(currentCRD.GetResourceVersion())
+					if len(matchedCSV) == 1 {
+						logger.Debugf("Found one owner for CRD %v", crd)
+					} else if len(matchedCSV) > 1 {
+						logger.Debugf("Found multiple owners for CRD %v", crd)
+
+						err := EnsureV1Beta1CRDVersions(currentCRD, crd)
+						if err != nil {
+							return v1alpha1.StepStatusUnknown, errorwrap.Wrapf(err, "error missing existing CRD version(s) in new CRD: %s", step.Resource.Name)
+						}
+
+						if err = validateV1Beta1CRDCompatibility(b.dynamicClient, currentCRD, crd); err != nil {
+							return v1alpha1.StepStatusUnknown, errorwrap.Wrapf(err, "error validating existing CRs agains new CRD's schema: %s", step.Resource.Name)
+						}
+					}
+					// Remove deprecated version in CRD storedVersions
+					storeVersions := removeDeprecatedV1Beta1StoredVersions(currentCRD, crd)
```

Comment:
this whole block is dangerous and might cause serious data loss. Just because a version not served anymore, does not mean that it is not stored in etcd anymore. If I follow you here, you are forcefully removing the remove version from the `StoredVersions` field. I.e. you are claiming that you have migrated every CR. Have you? I don't think.

Reply:
You are correct and we have discovered this code block to be faulty - we have removed it entirely from the code path and will address in a followup